src/iop/dual_demosaic.c

```diff
@@ -0,0 +1,255 @@
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+   Dual demosaicing has been implemented by Ingo Weyrich <heckflosse67@gmx.de> for
+   rawtherapee under GNU General Public License Version 3
+   and has been modified to work for darktable by Hanno Schwalm (hanno@schwalm-bremen.de).
+   Also the code for fast_blur has been taken from rawtherapee capturesharpening,
+   implemented also by Ingo Weyrich.
+*/
+#ifdef __GNUC__
+  #pragma GCC push_options
+  #pragma GCC optimize ("-Ofast")
+#endif
+
+static INLINE float calcBlendFactor(float val, float threshold)
+{
+    // sigmoid function
+    // result is in ]0;1] range
+    // inflexion point is at (x, y) (threshold, 0.5)
+    return 1.0f / (1.0f + dt_fast_expf(16.0f - (16.0f / threshold) * val));
+}
+
+static void fast_blur(float *const restrict src, float *const restrict out, const int width, const int height, const float sigma)
+{
+  // For a blurring sigma of 2.0f a 13x13 kernel would be optimally required but the 9x9 is by far good enough here 
+  float kernel[9][9];
+  const double temp = -2.0f * sqrf(sigma);
+  float sum = 0.0f;
+  for(int i = -4; i <= 4; i++)
+  {
+    for(int j = -4; j <= 4; j++)
+    {
+      if( (sqrf(i) + sqrf(j)) <= sqrf(3.0 * 1.5) )
+      {
+        kernel[i + 4][j + 4] = expf( (sqrf(i) + sqrf(j)) / temp);
+        sum += kernel[i + 4][j + 4];
+      }
+      else
+        kernel[i + 4][j + 4] = 0.f;
+    }
+  }
+  for(int i = 0; i < 9; i++)
+  {
+    for(int j = 0; j < 9; j++)
+      kernel[i][j] /= sum;
+  }
+  const float c42 = kernel[0][2];
+  const float c41 = kernel[0][3];
+  const float c40 = kernel[0][4];
+  const float c33 = kernel[1][1];
+  const float c32 = kernel[1][2];
+  const float c31 = kernel[1][3];
+  const float c30 = kernel[1][4];
+  const float c22 = kernel[2][2];
+  const float c21 = kernel[2][3];
+  const float c20 = kernel[2][4];
+  const float c11 = kernel[3][3];
+  const float c10 = kernel[3][4];
+  const float c00 = kernel[4][4];
+  const int w1 = width;
+  const int w2 = 2*width;
+  const int w3 = 3*width;
+  const int w4 = 4*width;
+#ifdef _OPENMP
+  #pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(src, out) \
+  shared(c42, c41, c40, c33, c32, c31, c30, c22, c21, c20, c11, c10, c00, w1, w2, w3, w4, width, height) \
+  schedule(simd:static) aligned(src, out : 64) 
+ #endif
+  for(int row = 4; row < height - 4; row++)
+  {
+#if defined(__clang__)
+        #pragma clang loop vectorize(assume_safety)
+#elif defined(__GNUC__)
+        #pragma GCC ivdep
+#endif
+    for(int col = 4; col < width - 4; col++)
+    {
+      const int i = row * width + col;
+      const float val = c42 * ((src[i - w4 - 2] + src[i - w4 + 2]) + (src[i - w2 - 4] + src[i - w2 + 4]) + (src[i + w2 - 4] + src[i + w2 + 4]) + (src[i + w4 - 2] + src[i + w4 + 2])) +
```

Comment:
In RT code I grouped the additions
```
((src[i - w4 - 2] + src[i - w4 + 2]) + (src[i - w2 - 4] + src[i - w2 + 4]) ....
```
to use more add slots. As you build with -Ofast you could try to remove the groupings. Maybe the compiler does a better job...

Reply:
Tested here on gcc 10.2.1 and indeed the compiler knows better ...